### PR TITLE
Fix incorrect version reporting in changelogs

### PR DESCRIPTION
**Pull request by cidrblock**
By default, RTD does a fetch --limit 50, if a tag is not present in the last 50 commits, the version reported by setuptools_scm will be incorrect and appears as ``v0.1.dev...`` in the towncrier changelog.

Another approach is to enable ``DONT_SHALLOW_CLONE`` for the repo

https://docs.readthedocs.io/en/stable/feature-flags.html#feature-flags
This was done for ansible-navigator on the day of this commit.

The changes force a `git fetch --unshallow` https://github.com/pypa/setuptools_scm/blob/7950bdaa3e5d9da42434f3e7a85d05971ebeca1f/src/setuptools_scm/git.py#L110

### CODE REVIEW

**Comment by webknjaz** [comment]
@cidrblock look at https://ansible-navigator--1001.org.readthedocs.build/en/1001/changelog/ — it does not have the tags fetched. IIRC this is fixable via the RTD config and needs to be included.

Also, the tox env needs to be changed to always recreate the tox virtualenv so that locally the version is updated.

File: .readthedocs.yml [hunk 1]
```diff
@@ -1,15 +1,12 @@
 # https://docs.readthedocs.io/en/stable/config-file/v2.html
-
```
Comment (webknjaz) [inline .readthedocs.yml:2]: Maybe avoid including formatting changes?

File: .readthedocs.yml [hunk 2]
```diff
@@ -20,6 +17,8 @@ formats: all
 
 python:
   install:
-  - method: setuptools
-    path: .
-  - requirements: docs/requirements.in
+    - method: pip
+      path: .
+    - method: setuptools
+      path: .
```
Comment (webknjaz) [inline .readthedocs.yml]: This shouldn't be necessary. `pip` will build and install it in the previous step. And at that point, all of the version metadata must already be present in the Git checkout.

Comment (webknjaz) [reply .readthedocs.yml]: Ah, wrong diff.

File: .readthedocs.yml [hunk 3]
```diff
@@ -1,9 +1,12 @@
 # https://docs.readthedocs.io/en/stable/config-file/v2.html
-
 ---
-
 version: 2
 
+build:
```
Comment (webknjaz) [inline .readthedocs.yml:5]: Could you submit this in a separate PR?

File: docs/conf.py [hunk 4]
```diff
@@ -36,16 +35,27 @@
...
-    get_scm_version(
-        local_scheme="no-local-version",
-    ).split(".")[:3],
-)
 
-# The full version, including alpha/beta/rc tags
-release = get_scm_version()
+# on_rtd is whether we are on readthedocs.org, this line of code grabbed from docs.readthedocs.org
+# https://docs.readthedocs.io/en/latest/faq.html?highlight=environ#how-do-i-change-behavior-for-read-the-docs
+on_rtd = os.environ.get("READTHEDOCS", None) == "True"
```
Comment (webknjaz) [inline docs/conf.py:41]: This shouldn't be necessary. Aim for having the correct metadata in the package instead. I think I solved that with https://github.com/ansible/pylibssh/blob/0610b59/.readthedocs.yml#L28 in the past.

File: setup.py [hunk 5]
```diff
@@ -5,6 +5,7 @@
 
 import glob
 import os
+import shutil
```
Comment (webknjaz) [inline setup.py:8]: This seems forgotten.
```suggestion
```

File: tox.ini [hunk 6]
```diff
@@ -134,6 +134,7 @@ commands =
 isolated_build = true
 passenv =
   SSH_AUTH_SOCK
+recreate = true
```
Comment (webknjaz) [inline tox.ini:137]: (educational) Note that this will guarantee an up-to-date version in the metadata at the expense of populating a fresh virtualenv on each execution.

File: docs/conf.py [hunk 7]
```diff
@@ -3,17 +3,16 @@
...
 """Configuration file for the Sphinx docs."""
 
-from functools import partial
+import os
+
+from importlib.metadata import version as get_version
 from pathlib import Path
 from sys import path
 
-from setuptools_scm import get_version
```
Comment (webknjaz) [inline docs/conf.py:10]: What is the reason you don't want to keep this? I think it's rather reasonable to have this as a docs dependency.

File: docs/conf.py [hunk 8]
```diff
@@ -36,16 +35,27 @@
...
-)
 
-# The full version, including alpha/beta/rc tags
-release = get_scm_version()
+# on_rtd is whether we are on readthedocs.org, this line of code grabbed from docs.readthedocs.org
+# https://docs.readthedocs.io/en/latest/faq.html?highlight=environ#how-do-i-change-behavior-for-read-the-docs
+on_rtd = os.environ.get("READTHEDOCS", None) == "True"
+if on_rtd:
+    rtd_version = os.environ["READTHEDOCS_VERSION"]
+    release = rtd_version
```
Comment (webknjaz) [inline docs/conf.py:44]: Looks like one of the vars needs to be converted to a short version while the other should preserve the full one. The problem with this is that RTD does not know about those local and dev version details that setuptools-scm extracts from Git. So it will probably make it impossible to replicate the behavior exactly.

Comment (webknjaz) [reply docs/conf.py:44]: I think the commit hash is somehow exposed on RTD but it's still not the same — the version base that is supposed to be calculated from Git for intermediate (in-between-the-tags) commits is nowhere to be found.

**Review by webknjaz — CHANGES_REQUESTED** [review]
> This is the suggested approach in the setuptools_scm project: [pypa/setuptools_scm#usage-from-sphinx](https://github.com/pypa/setuptools_scm#usage-from-sphinx)

The more I look into this the more I realize (and start remembering) that the suggested approach does not cover the use-cases we've got. The main one being the Towncrier drafts preview integration.

So if the only motivation behind is change is "because the README only has this example", I think that breaking the integration I've polished while implementing it across many projects is not worth it.

> This should also get us one step closer to removing the ignore_import flag for mypy

What is this about?

That said, I don't see how the suggested approach can realistically work for us. If there's some annoyance with typing, maybe just add an inline type-ignore for now and drop the rest?

**Comment by cidrblock** [comment]
> What is this about?

Originally, yes, simply avoid a type ignore[import], but it doesn't appear this has ever worked correctly. So now it's a testing ground to see if it can be fixed.

File: tox.ini (hunk 6 above)
Comment (cidrblock) [reply tox.ini:137]: "Also, the tox env needs to be changed to always recreate the tox virtualenv so that locally the version is updated."

I think I misunderstood the ask here...

File: docs/conf.py [hunk 9]
```diff
@@ -8,12 +8,36 @@
...
+    present in the last 50 commits, the version reported by setuptools_scm
+    will be incorrect and appears as ``v0.1.dev...`` in the towncrier changelog.
+    Another approach is to enable ``DONT_SHALLOW_CLONE`` for the repo
+    https://docs.readthedocs.io/en/stable/feature-flags.html#feature-flags
+    This was done for ansible-navigator on the day of this commit.
+
+    :param args: The arguments
+    :param kwargs: The keyword arguments
+    :returns: The parsed version
+    """
```
Comment (webknjaz) [inline docs/conf.py:31]: Since it's not a lambda, we can properly declare the expectations here now:
```suggestion
    """
    assert 'pre_parse' not in kwargs
```

File: docs/conf.py [hunk 10]
```diff
@@ -8,12 +8,36 @@
...
+
+def parse_with_fetch(*args, **kwargs) -> str:
+    """If the repo is found to be shallow, fetch a full.
+
+    By default, RTD does a fetch --limit 50, if a tag is not
+    present in the last 50 commits, the version reported by setuptools_scm
+    will be incorrect and appears as ``v0.1.dev...`` in the towncrier changelog.
+    Another approach is to enable ``DONT_SHALLOW_CLONE`` for the repo
+    https://docs.readthedocs.io/en/stable/feature-flags.html#feature-flags
+    This was done for ansible-navigator on the day of this commit.
```
Comment (webknjaz) [inline docs/conf.py:26]: I think this part should probably be in a code comment. WDYT?

File: docs/conf.py [hunk 11]
```diff
@@ -8,12 +8,36 @@
...
+
+    :param args: The arguments
+    :param kwargs: The keyword arguments
+    :returns: The parsed version
+    """
+    return parse(*args, pre_parse=fetch_on_shallow, **kwargs)
+
+
+get_scm_version = partial(get_version, parse=parse_with_fetch)
```
Comment (webknjaz) [inline docs/conf.py:35]: Do we want to have this conditional for just RTD?

File: tox.ini [hunk 12]
```diff
@@ -103,6 +103,9 @@ deps =
   # -c{toxinidir}/docs/requirements.txt
 description = Build The Docs
 changedir = {toxinidir}/docs
+# Reinstall so setuptools-scm runs and updates the _version.py
+# intentionally not in alphabetical order due to execution order
+commands_pre = {envpython} -m pip install --no-deps -e {toxinidir}
```
Comment (webknjaz) [inline tox.ini:108]: Also, given that the current implementation does not use the dist metadata, this isn't necessary, strictly speaking.

File: tox.ini [hunk 13]
```diff
@@ -103,6 +103,9 @@ deps =
...
 commands =
   # Retrieve possibly missing commits:
   -git fetch --unshallow
```
Comment (webknjaz) [inline tox.ini:111]: With the same code in `conf.py`, this could be dropped given that this operation remains unconditional there.

**Review by webknjaz — COMMENTED** [review]
So here's a few ideas to check out before merging.

File: docs/conf.py (hunk 9 above)
Comment (cidrblock) [reply docs/conf.py:31]: interesting..........I guess good insurance

File: docs/conf.py (hunk 11 above)
Comment (cidrblock) [reply docs/conf.py:36]: I was thinking what if someone somehow managed to do this locally....what would the harm be, also, if the experimental func ever changes or acts up, it should be able to repro locally w/o having to set the magic RTD envvar

File: tox.ini (hunk 12 above)
Comment (cidrblock) [reply tox.ini]: True, but the tox env not being up to date seems odd, what I really should add is `git clean -iX -- docs/pkg/` because I've been bitten by that with failed tox docs builds a bunch

File: tox.ini (hunk 13 above)
Comment (cidrblock) [reply tox.ini:111]: right!

Comment (cidrblock) [reply tox.ini:111]: know if this will gets the tags too?
```
--unshallow
If the source repository is complete, convert a shallow repository to a complete one, removing all the limitations imposed by shallow repositories.

If the source repository is shallow, fetch as much as possible so that the current repository has the same history as the source repository.
```

File: docs/conf.py (hunk 10 above)
Comment (cidrblock) [reply docs/conf.py:26]: I think the explanation is helpful in the docstring since it's such an odd thing
image

Comment (webknjaz) [reply docs/conf.py:26]: Ah, okay. Usually, the motivation (the "why") goes into comments but the "how"/"what" goes into docstrings.

File: docs/conf.py (hunk 11 above)
Comment (webknjaz) [reply docs/conf.py:36]: Agreed. Just making sure all the questions are asked and answered :)

File: docs/conf.py (hunk 9 above)
Comment (webknjaz) [reply docs/conf.py:31]: Not just insurance but also expectation declaration per contract programming principles :)

**Review by webknjaz — APPROVED** [review]
Brilliant! Thanks for taking care of this!

File: tox.ini (hunk 12 above)
Comment (webknjaz) [reply tox.ini]: True, that is a great candidate for the next PR :)

File: tox.ini (hunk 13 above)
Comment (webknjaz) [reply tox.ini]: Not sure. I've been using the two commands for years so maybe it wasn't true at the time I first introduced it.

**Comment by cidrblock** [comment]
> Brilliant! Thanks for taking care of this!

Couldn't have done it w/o your help.
